### PR TITLE
[PodTarget] Infer umbrella header

### PR DESCRIPTION
**Pull request by segiddins**
Might help with https://github.com/CocoaPods/CocoaPods/issues/4420?
Closes https://github.com/CocoaPods/CocoaPods/issues/3767

### CODE REVIEW

**Comment by segiddins** [comment]
Note that, with this change, spec authors will have to make their umbrella headers conditional if they have subspecs (i.e. checking `#if __has_include`)

**Comment by segiddins** [comment]
(The other alternative might be to append imports of all public headers to the existing header file?)

**Comment by segiddins** [comment]
@CocoaPods/core what are the thoughts on this one?

**Comment by AliSoftware** [comment]
This breaks `OHHTTPStubs`: `pod lib lint` returns errors, while it passes with the current 0.39).

```
 -> OHHTTPStubs (4.5.0)
    - WARN  | [OHHTTPStubs/Default,OHHTTPStubs/JSON] xcodebuild:  <module-includes>:1:1: warning: umbrella header for module 'OHHTTPStubs' does not include header 'OHHTTPStubsResponse+JSON.h' [-Wincomplete-umbrella]
    - WARN  | [OHHTTPStubs/Default] xcodebuild:  <module-includes>:1:1: warning: umbrella header for module 'OHHTTPStubs' does not include header 'OHPathHelpers.h' [-Wincomplete-umbrella]
    - ERROR | [OHHTTPStubs/Default, OHHTTPStubs/Core, OHHTTPStubs/NSURLSession, and more...] xcodebuild: Returned an unsuccessful exit code.
    - WARN  | [OHHTTPStubs/HTTPMessage] xcodebuild:  <module-includes>:1:1: warning: umbrella header for module 'OHHTTPStubs' does not include header 'OHHTTPStubsResponse+HTTPMessage.h' [-Wincomplete-umbrella]
    - WARN  | [OHHTTPStubs/Mocktail] xcodebuild:  <module-includes>:1:1: warning: umbrella header for module 'OHHTTPStubs' does not include header 'OHHTTPStubs+Mocktail.h' [-Wincomplete-umbrella]

[!] OHHTTPStubs did not pass validation, due to 1 error and 4 warnings.
```
[Full log of `pod lib lint --verbose` here](https://github.com/CocoaPods/CocoaPods/files/24832/lint-OHHTTPStubs-infer-umbrella-header.txt)

---

Will try to investigate later, but have to go and will be busy until 17pm Paris time. Maybe I'm just missing some `__has_include` stuff at some point (I had them but not sure I kept them when migrating to a custom umbrella header), will have to double-check that.

Note that `OHHTTPStubs`:
* Has subspecs
* Has a custom umbrella header and module map
* Is iOS/watchOS/tvOS/OSX and has been successfully tested to work on Xcode 7.1 before.
* Umbrella header issue was recently fixed via AliSoftware/OHHTTPStubs#127

File: lib/cocoapods/target.rb [hunk 1]
```diff
@@ -140,14 +140,14 @@ def xcconfig_path(variant = nil)
     #         module map.
     #
     def umbrella_header_path
-      support_files_dir + "#{label}-umbrella.h"
+      support_files_dir + "#{product_module_name}.h"
```
Comment (AliSoftware) [inline lib/cocoapods/target.rb]: Won't this change only work if the umbrella header is the inferred one, but break if the module map has a custom umbrella header instead (like in `OHHTTPStubs` where `OHHTTPStubs.h` exists but **isn't** the umbrella header)?

**Comment by AliSoftware** [comment]
I think this change will do more harm than solve stuff, because:

* People will have to add `__has_include` everywhere in their umbrella headers, whereas they didn't have to care before as CocoaPods did it right
* This inferred umbrella header can guess wrong (like in the case of `OHHTTPStubs` if the header with the same name already exists but isn't the umbrella header (which is probably the most common reason why people would do a custom module map in the first place)

Wouldn't a `custom_module_map` DSL attribute help fix that? People wanting CocoaPods to use their own modulemap & umbrella header instead of letting CocoaPods generate its own could explicitly declare it in their `.podspec` that way, instead of guessing and risking to break things?

**Comment by segiddins** [comment]
@AliSoftware if you have a custom module map, then CocoaPods already doesn't generate an umbrella header.

File: lib/cocoapods/target.rb (hunk 1 above)
Comment (segiddins) [inline lib/cocoapods/target.rb]: this code path already isn't used when a custom module map is specified

**Comment by AliSoftware** [comment]
Mmmh how come does it still work if a user only fetch some subspecs and not all, then?

Note that I didn't mention anything my `podspec` about that custom umbrella header, because:
* if people are using my lib thru anything other than CocoaPods, like Carthage or manual copy of the framework or whatnot, they'll have all the subspecs in that generated framework and my custom umbrella header will import them all
* if people are using my lib thru CocoaPods, I let it generate the umbrella header for me so that it only import the headers that are actually fetched by the subspecs selected in the Podfile

You're saying that there's a DLS attribute to specify a custom module map already, and that I should use it instead so that CocoaPods _never_ generate its own module map in my case? (And thus I'll need to add `__has_include` tests in that custom umbrella headers so it can work with subspecs then)?

**Comment by segiddins** [comment]
I'm saying if you're using a custom module map, CocoaPods will not generate an umbrella header.

**Pull request closed without merging**